# <center>Random Forest Classifier</center>
Random forests or random decision forests are an ensemble learning method for classification, regression and other tasks, that operate by constructing a multitude of decision trees at training time and outputting the class that is the mode of the classes (classification) or mean prediction (regression) of the individual trees. Random decision forests correct for decision trees’ habit of over fitting to their training set.

![](img/img-DecisionTree_RandomForestClassifier/random_for.jpg)
![](img/img-DecisionTree_RandomForestClassifier/random_forest.png)

## Ensemble models:

It is a model that internally uses group of models to make prediction.

In sklearn we have multiple ensemble models

**RandomForest:**
- Group of decision trees
    
**Boosting Models:**
- AdaBoost
- GradientBoost
- etc
    
**VotingClassifier:**
- Group of any type of models 

In [8]:
import pandas as pd
df=pd.read_csv('B:/dataset/weather_tree.csv')
df.head()

Outlook  Temp Humidity  Windy Play
0   Rainy   Hot     High  False   No
1   Rainy   Hot     High   True   No
2   Rainy  Mild     High  False   No
3   Rainy  Cool   Normal  False  Yes
4   Rainy  Mild   Normal   True  Yes

In [9]:
df['Outlook']=df.Outlook.map({'Rainy':0,'Overcast':1,'Sunny':2})
df['Temp']=df.Temp.map({'Hot':0,'Mild':1,'Cool':2})
df['Humidity']=df.Humidity.map({'High':0,'Normal':1})

In [10]:
df.head()

Outlook  Temp  Humidity  Windy Play
0        0     0         0  False   No
1        0     0         0   True   No
2        0     1         0  False   No
3        0     2         1  False  Yes
4        0     1         1   True  Yes

In [11]:
X=df.iloc[:,:-1]
y=df.iloc[:,-1].values

In [12]:
from sklearn.tree import DecisionTreeClassifier
dtree=DecisionTreeClassifier(criterion='gini')
dtree.fit(X,y)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [13]:
pred=dtree.predict(X)

In [14]:
from sklearn.metrics import accuracy_score
accuracy_score(y,pred)

1.0

In [15]:
test_sample=[['Overcast','Hot','Normal',False],['Rainy','Hot','High',False]]

In [16]:
test_sample=[[1,0,1,False],[0,0,0,False]]

In [17]:
dtree.predict(test_sample)

array(['Yes', 'No'], dtype=object)

In [18]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier(n_estimators=3)
rf.fit(X,y)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=3, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

In [19]:
rf.predict(test_sample)

array(['Yes', 'No'], dtype=object)

## FutureWarning

In [1]:
# example of RandomForestClassifier that generates a FutureWarning
from sklearn.datasets import make_blobs
from sklearn.ensemble import RandomForestClassifier
# prepare dataset
X, y = make_blobs(n_samples=100, centers=2, n_features=2)
# create and configure model
model = RandomForestClassifier()
# fit model
model.fit(X, y)

D:\Anaconda\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

**D:\Anaconda\lib\site-packages\sklearn\ensemble\forest.py:246: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)**
  
# create and configure model
model = RandomForestClassifier(n_estimators=10)